---
title: "Metagenome-atlas output summary"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = TRUE, message = FALSE, 
                      warning = FALSE, cache = FALSE, include = TRUE)
knitr::opts_knit$set(root.dir = "..") # set root dir one level above where Rmd is saved
```

Atlas output files are stored in the `Example` folder in the GitHub repository `metagenome-atlas/Tutorial`.

If you don't already have a copy of the GitHub repository locally, you can run the following command from a Terminal. 
Note that RStudio includes a `Terminal` tab next to the `Console` tab. 
In the default configuration of RStudio, this panel is located in the bottom left hand corner.

```{r, eval = F}
git clone https://github.com/metagenome-atlas/Tutorial.git
```

```{r libraries}
library(dplyr)
library(readr)
library(stringr)
library(tidyr)
library(tibble)
library(ggplot2)
library(plotly)
library(pheatmap)
library(microbiome)
library(ape)
library(vegan)
library(kableExtra)
```

# Metagenome assembled genomes

## Taxonomy

```{r read_tax}
Tax <- read_tsv("Example/Results/taxonomy.tsv")
kable(Tax)
```

```{r lab_tax}
# create a short label for each species
Tax <- Tax %>%
  mutate(Labels = ifelse(is.na(species) & is.na(genus), paste0(family, " ", user_genome), species)) %>%
  mutate(Labels = ifelse(is.na(Labels), paste0(genus, " ", user_genome), Labels))
```

## Genome quality

```{r read_genome_quality}
genome_quality <- read_tsv("Example/Results/genome_completeness.tsv") %>%
  mutate(Quality_Score = Completeness - (5*Contamination)) %>%
  mutate(Lineage = sapply(str_split(`Marker lineage`, " "), function(x) x[1])) %>%
  left_join(Tax, by = c("Bin Id" = "user_genome")) %>%
  mutate(Name = Labels) %>%
  select(-Labels)
```

```{r plot_contam_vs_quality}
plt <- ggplot(genome_quality, aes(x = Contamination, y = Completeness, color = phylum)) +
  geom_point() +
  theme_minimal()
ggplotly(plt)
```

## Abundance

```{r read_counts}
Counts <- read_tsv("Example/Results/counts/raw_counts_genomes.tsv")
kable(head(Counts))
```

### Mapping rate

```{r mapping_rate}
mapping_rate <- read_tsv("Example/Results/mapping_rate.tsv", 
                         col_names = c("Sample", "Mapping_rate"), skip = 1)  %>%
  mutate(tmp = "samples")

ggplot(mapping_rate, aes(x = tmp, y = Mapping_rate)) +
  geom_jitter() +
  ylim(c(0, 1))+
  theme_minimal()
```

### Stats based on raw counts

There are good reasons to use rawcounts and use centric log ratios, see more in 

Gloor, Gregory B., Jean M. Macklaim, Vera Pawlowsky-Glahn, and Juan J. Egozcue. 2017. “Microbiome Datasets Are Compositional: And This Is Not Optional.” Frontiers in Microbiology 8 (November). Frontiers: 2224. doi:10.3389/fmicb.2017.02224.


For differencial abundance analysis see also the same paper.

```{r clr_transform}
# transform counts with centered log ratio
data <- Counts %>%
  column_to_rownames(var = "Sample") %>%
  as.data.frame()
data <- transform(data, "clr")
```

```{r heatmap}
pheatmap(data, cluster_rows = T, cluster_cols = T,
         labels_row = Tax$Labels)
```

### PCA (PCoA) of the robust aitchison distance

```{r pcoa_aitchison}
pcoa_data <- pcoa(dist(t(data), method = "euclidean"))
pcoa_df <- as.data.frame(pcoa_data$vectors)

plt <- ggplot(pcoa_df, aes(x = Axis.1, y = Axis.2, label = rownames(pcoa_df))) +
  geom_point() +
  labs(x = paste0("PC1 (", round(pcoa_data$values$Relative_eig[1] * 100, digits = 1), "%)"),
       y = paste0("PC2 (", round(pcoa_data$values$Relative_eig[2] * 100, digits = 1), "%)")) +
  theme_minimal()

ggplotly(plt)
```

### Relative abundance

For the relative abundance we take the coverage over the genome not the raw counts. This inmplicit normalizes for genome size. The coverage is calculated as the median of the coverage values calculated in 1kb blocks.

```{r}
D <- read_tsv("Example/Results/counts/median_coverage_genomes.tsv") %>%
  column_to_rownames(var = "X1") %>%
  as.data.frame()
kable(head(D))
```

```{r relative_abund}
# calculate relative abundance
rel_ab <- sweep(D, 1, rowSums(D),`/`)
```

```{r calc_most_abund}
# get most abundant genomes
counts_per_genome <- data.frame(sums = colSums(rel_ab)) %>%
  rownames_to_column(var = "Sample") %>%
  left_join(Tax, by = c("Sample" = "user_genome")) %>%
  arrange(desc(sums))
  

ggplot(counts_per_genome %>%
         top_n(sums, n = 10), aes(x = reorder(Labels, sums), y = sums)) +
  geom_col() +
  labs(x = "", y = "Abundance [rel_ab]", title = "Most abundant genomes") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90))
```

### Typical bar chart

```{r barchart}
level <- 'family'

grouped_data <- rel_ab %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "user_genome") %>%
  pivot_longer(cols = -user_genome, names_to = "Sample", values_to = "rel_ab") %>%
  left_join(Tax, by = "user_genome") %>%
  group_by(Sample, family) %>%
  summarise(summarized_rel_ab = sum(rel_ab))

ggplot(grouped_data, aes(x = Sample, y = summarized_rel_ab, fill = family)) +
  geom_col() +
  theme_minimal() + 
  theme(axis.text.x = element_text(angle = 90))
```

## Functional annotation

Relative abundance of functional annotations per sample

The abundance is calculated as the sum of the relative abundance of all bacteria containing a function.

### CAZy -- Carbohydrate active enzymes

```{r read_cazy}
CAZy_annotations_genome <- read_tsv('Example/Results/annotations/CAZy.tsv')
CAZy_presence <- CAZy_annotations_genome %>%
  column_to_rownames("MAG")
CAZy_presence[CAZy_presence > 0] <- 1
kable(head(CAZy_presence))


CAZy_rel_ab <- as.matrix(rel_ab) %*% as.matrix(CAZy_presence)
kable(head(CAZy_rel_ab))
pheatmap(CAZy_rel_ab)
```

### KEGG orthologs

```{r read_kegg}
Kegg_annotations_genome <- read_tsv('Example/Results/annotations/KO.tsv')
Kegg_presence <- Kegg_annotations_genome %>%
  column_to_rownames("MAG")
Kegg_presence[Kegg_presence > 0] <- 1
kable(head(Kegg_presence))


Kegg_rel_ab <- as.matrix(rel_ab) %*% as.matrix(Kegg_presence)
kable(head(Kegg_rel_ab))
pheatmap(Kegg_rel_ab)
```
