---
title: "Analyze the output of metagenome-atlas"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = TRUE, message = FALSE,
                      warning = FALSE, cache = FALSE, include = TRUE)
knitr::opts_knit$set(root.dir = "..") # set root dir one level above where Rmd is saved
```

Atlas output files are stored in the `Example` folder.

```{r libraries}
library(dplyr)
library(readr)
library(stringr)
library(tidyr)
library(tibble)
library(ggplot2)
library(plotly)
library(pheatmap)
library(microbiome)
library(ape)
library(vegan)
install.packages("useful")
library(useful)
library(kableExtra)
library(ggtree)
```

# Taxonomy

```{r read_tax}
Tax <- read_tsv("Example/Results/taxonomy.tsv")
kable(Tax)
```

```{r lab_tax}
# create a short label for each species
Tax <- Tax %>%
  mutate(Labels = ifelse(is.na(species) & is.na(genus), paste0(family, " ", user_genome), species)) %>%
  mutate(Labels = ifelse(is.na(Labels), paste0(genus, " ", user_genome), Labels))
```
## Draw tree

```{r read_newick}
T <- ggtree::read.tree("Example/genomes/tree/gtdbtk.bac120.nwk")
```

```{r tax_labels}
# node names are stored in T$tip.label.
Tax <- Tax[order(match(Tax$user_genome, T$tip.label)), ] # sort Tax according to the tree
stopifnot(all.equal(Tax$user_genome, T$tip.label)) # make sure the sorting worked
T$tip.label <- Tax$Labels
```

```{r color_tree, fig.height=10}
# color tree by phylum
ggtree(T, layout = 'circular') %<+% Tax +
  geom_tiplab(aes(color = phylum)) +
  theme(legend.position = "bottom")
```


```{r highlight_tree, fig.height = 10}
library(phytools)
Bacteroidota <- findMRCA(T, c("RC9 sp002298075", "Paramuribaculum MAG61"))

ggtree(T, layout = 'circular') +
  geom_tiplab() +
  geom_highlight(Bacteroidota, fill = "steelblue", extend = 3)
```



# Genome quality

```{r read_genome_quality}
genome_quality <- read_tsv("Example/Results/genome_completeness.tsv") %>%
  mutate(Quality_Score = Completeness - (5*Contamination)) %>%
  mutate(Lineage = sapply(str_split(`Marker lineage`, " "), function(x) x[1])) %>%
  left_join(Tax, by = c("Bin Id" = "user_genome")) %>%
  mutate(Name = Labels) %>%
  select(-Labels)
```

```{r plot_contam_vs_quality}
plt <- ggplot(genome_quality, aes(x = Contamination, y = Completeness, color = phylum)) +
  geom_point() + xlab('Contamination [%]') + ylab('Completeness [%]') +
  theme_minimal()
ggplotly(plt)
```

# Abundance

```{r read_counts}
Counts <- read_tsv("Example/Results/counts/raw_counts_genomes.tsv")
kable(topleft(Counts, c = 10))
```

## Mapping rate

```{r mapping_rate, fig.width=2, fig.height=4}
mapping_rate <- read_tsv("Example/Results/mapping_rate.tsv",
                         col_names = c("Sample", "Mapping_rate"), skip = 1)  %>%
  mutate(tmp = "samples")

ggplot(mapping_rate, aes(x = tmp, y = Mapping_rate)) +
  geom_jitter() +
  ylim(c(0, 1))+
  theme_minimal()
```


## Relative abundance

For the relative abundance, we take the coverage over the genome, not the raw counts. This implicitly normalizes for genome size. The coverage is calculated as the median of the coverage values calculated in 1kb blocks.

```{r}
D <- read_tsv("Example/Results/counts/median_coverage_genomes.tsv") %>%
  column_to_rownames(var = "X1") %>%
  as.data.frame()
kable(topleft(D, c= 10))
```

```{r relative_abund}
# calculate relative abundance
rel_ab <- sweep(D, 1, rowSums(D),`/`)
```

```{r calc_most_abund}
# get most abundant genomes
counts_per_genome <- data.frame(sums = colSums(rel_ab)) %>%
  rownames_to_column(var = "Sample") %>%
  left_join(Tax, by = c("Sample" = "user_genome")) %>%
  arrange(desc(sums))


ggplot(counts_per_genome %>%
         top_n(sums, n = 10), aes(x = reorder(Labels, sums), y = sums)) +
  geom_col() +
  labs(x = "", y = "Abundance [rel_ab]", title = "Most abundant genomes") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90))
```

### Typical bar chart

```{r barchart}
level <- 'family'

grouped_data <- rel_ab %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "user_genome") %>%
  pivot_longer(cols = -user_genome, names_to = "Sample", values_to = "rel_ab") %>%
  left_join(Tax, by = "user_genome") %>%
  group_by(Sample, family) %>%
  summarise(summarized_rel_ab = sum(rel_ab))

ggplot(grouped_data, aes(x = Sample, y = summarized_rel_ab, fill = family)) +
  geom_col() +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90)) +
  #scale_fill_manual(values=rep(brewer.pal(5,"Paired"), times=5))
  scale_fill_manual(values = c("#771155", "#AA4488", "#CC99BB", "#114477", "#4477AA", "#77AADD", "#117777", "#44AAAA", "#77CCCC", "#117744", "#44AA77", "#88CCAA", "#777711", "#AAAA44", "#DDDD77", "#774411", "#AA7744", "#DDAA77", "#771122", "#AA4455", "#DD7788"))

```

# Functional annotation

The relative abundance of functional annotations per sample

The abundance is calculated as the sum of the relative abundance of all bacteria containing a function.

## CAZy -- Carbohydrate active enzymes

```{r read_cazy}
CAZy_annotations_genome <- read_tsv('Example/Results/annotations/CAZy.tsv')
CAZy_presence <- CAZy_annotations_genome %>%
  column_to_rownames("MAG")
CAZy_presence[CAZy_presence > 0] <- 1
kable(topleft(CAZy_presence, c=10))


CAZy_rel_ab <- as.matrix(rel_ab) %*% as.matrix(CAZy_presence)
kable(topleft(CAZy_rel_ab, c=10 ))
pheatmap(CAZy_rel_ab, show_colnames = F)
```

## KEGG orthologs

```{r read_kegg}
Kegg_annotations_genome <- read_tsv('Example/Results/annotations/KO.tsv')
Kegg_presence <- Kegg_annotations_genome %>%
  column_to_rownames("MAG")
Kegg_presence[Kegg_presence > 0] <- 1
kable(topleft(Kegg_presence, c=10 ))


Kegg_rel_ab <- as.matrix(rel_ab) %*% as.matrix(Kegg_presence)
kable(topleft(Kegg_rel_ab, c=10 ))
pheatmap(Kegg_rel_ab, show_colnames = F)
```
