---
title: "Differential Abundance Analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = TRUE, message = FALSE,
                      warning = FALSE, cache = FALSE, include = TRUE)
```

```{r libraries,inlclude=FALSE}
library(dplyr)
library(readr)
library(tibble)
library(tidyr)
library(purrr)
library(broom)
library(pheatmap)
library(plotly)
library(microbiome)
library(ggbeeswarm)
library(knitr)
library(ALDEx2)

```

# How does high-fat diet impact the microbiome of mice?

![Picture of obese mice](https://upload.wikimedia.org/wikipedia/commons/0/0b/Fatmouse.jpg)

```{r, file_paths}
atlas_wd_folder = "../Example/"

taxonomy_file = paste0(atlas_wd_folder,"genomes/taxonomy/gtdb_taxonomy.tsv")
tree_file = paste0(atlas_wd_folder,"genomes/tree/gtdbtk.bac120.nwk")
quality_file= paste0(atlas_wd_folder,"genomes/checkm/completeness.tsv")
counts_file= paste0(atlas_wd_folder,"genomes/counts/raw_counts_genomes.tsv")
abundance_file = paste0(atlas_wd_folder,"genomes/counts/median_coverage_genomes.tsv")
readstats_file= paste0(atlas_wd_folder,"stats/read_counts.tsv")
keggmodules_file = paste0(atlas_wd_folder,"genomes/annotations/dram/kegg_modules.tsv")
```

```{r load_metadata}
# load metadata
metadata <- read_tsv('../Example/metadata.txt',show_col_types = FALSE) %>% column_to_rownames("...1")
kable(metadata)
```

We confirm that the mice on high-fat diet really put more weight on.

```{r plot_metadata, fig.width=1, fig.height=1.5}
ggplot(metadata, aes(x = Diet, y = Body_weight)) +
 geom_beeswarm(cex=5) +
  ylim(10,35)+
  theme_minimal()
```

```{r labels}
# create a short label for each genome
Tax <- read_tsv(taxonomy_file,show_col_types = FALSE) %>%
  mutate(Label = ifelse(is.na(species) & is.na(genus), paste0(family, " ", user_genome), species)) %>%
  mutate(Label = ifelse(is.na(Label), paste0(genus, " ", user_genome), Label))
```

## Relative abundance

For the relative abundance, we take the coverage over the genome, not the raw counts. This implicitly normalizes for genome size. The coverage is calculated as the median of the coverage values calculated in 1kb blocks.

```{r, relab}
D <- read_tsv(abundance_file, show_col_types = FALSE) %>%
  column_to_rownames(var = "...1") %>%
  as.data.frame()
# calculate relative abundance
rel_ab <- sweep(D, 1, rowSums(D),`/`)

# Subset metadata ther is one sample not quantified
metadata = metadata[rownames(rel_ab),]

```

### Bar chart wich group labels

```{r}
level <- 'phylum'

grouping_columns= c("Sample",level,"Diet")
grouping_columns <- lapply(grouping_columns, as.symbol)

grouped_data <- rel_ab %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column(var = "user_genome") %>%
  pivot_longer(cols = -user_genome, names_to = "Sample", values_to = "rel_ab") %>%
  left_join(Tax, by = "user_genome") %>%
  left_join(rownames_to_column(metadata), by = c("Sample" = "rowname")) %>%
  group_by_(.dots = grouping_columns) %>%
  summarise(summarized_rel_ab = sum(rel_ab))

ggplot(grouped_data, aes_string(x = "Sample", y = "summarized_rel_ab", fill = level)) +
  geom_col() +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90),
        axis.text.y = element_blank(),
        axis.title.y = element_blank()) +
  facet_wrap(~Diet, scales = "free_x") +
  scale_fill_manual(values = c("#771155", "#AA4488", "#CC99BB", "#114477", "#4477AA", "#77AADD", "#117777", "#44AAAA", "#77CCCC", "#117744", "#44AA77", "#88CCAA", "#777711", "#AAAA44", "#DDDD77", "#774411", "#AA7744", "#DDAA77", "#771122", "#AA4455", "#DD7788"))
```



## Compositional data analysis


In order to analyze the microbiome at the species or genome-level we use compositional data analysis (CoDa), see more on [Wikipedia](https://en.wikipedia.org/wiki/Compositional_data) and this article:

>Gloor, Gregory B., Jean M. Macklaim, Vera Pawlowsky-Glahn, and Juan J. Egozcue. 2017. “Microbiome Datasets Are  Compositional: And This Is Not Optional.” Frontiers in Microbiology 8 (November). Frontiers: 2224.
    doi: [10.3389/fmicb.2017.02224](https://www.frontiersin.org/articles/10.3389/fmicb.2017.02224).


For this, we transform the abundances with the centric log-ratios (CLR) after imputing values for the zeros.




```{r}
# transforme counts with centered log ratio
data <- transform(rel_ab, transform = "clr")
```

### PCA of the Aitchison distance

```{r pca}
transformed_data <- prcomp(data,scores=TRUE)

pca_data <- transformed_data$x %>%
  as.data.frame() %>%
  rownames_to_column(var = "Sample") %>%
  left_join(rownames_to_column(metadata), by = c("Sample" = "rowname"))

ggplot(pca_data, aes(x = PC1, y = PC2, color = Diet)) +
  geom_point() +
  theme_minimal() +
  scale_color_manual(values = c(chow = "#00BFC4", HF = "#F8766D"))
```



### Differential abundance analysis

As the counts are normalized in centred log-ratio the log FC becomes the difference.

We use the alde2 package that uses a welch test to assess differential abundance in the two groups. See Gloor et al for more information.

```{r aldex2}

Counts <- read_tsv(counts_file, show_col_types = FALSE) %>% column_to_rownames("Sample") %>% as.matrix() #%>% t()

x <- aldex.clr(reads= Counts, conds= metadata$Diet,mc.samples=16, denom="all", verbose=F)

d.eff <- aldex.effect(x)
d.tt <- aldex.ttest(x)

aldex_stats <- data.frame(d.eff,d.tt)

# rename
aldex_stats <- aldex_stats %>%
  rename(
    chow= rab.win.chow ,
    HF= rab.win.HF,
    logFC= diff.btw,
    Effect = effect ,
    Pvalue= we.ep ,
    Pvalue_BH = we.eBH
    ) %>%
  rownames_to_column(var = 'Id') %>%
  left_join(Tax, by= c("Id"= "user_genome")) %>%
  mutate(logP = -log10(Pvalue))


```

#### heatmap of significant Genomes

```{r heatmap, fig.height=3, fig.width=5}
# filter to MAG abundances that were significantly different
sig_data <- data[, aldex_stats[aldex_stats$Pvalue_BH < 0.1, ]$Id] %>%
  as.data.frame() %>%
  t()

# make a dataframe to use to annotate the heatmap
annot_df <- data.frame(Sample = colnames(sig_data)) %>%
  left_join(rownames_to_column(metadata), by = c("Sample" = "rowname")) %>%
  column_to_rownames(var = "Sample") %>%
  dplyr::select(Diet)

# sort labels by sig_data order
heatmap_labels <- Tax %>%
  filter(user_genome %in% rownames(sig_data))
heatmap_labels <- heatmap_labels[order(match(heatmap_labels$user_genome, rownames(sig_data))), ]

pheatmap(sig_data, annotation_col = annot_df,
         labels_row = heatmap_labels$Label)
```

#### Volcano plot

```{r}
## non interactive plot
# ggplot(aldex_stats, aes(x = logFC, y = logP, alpha = logP)) +
#  geom_point(color = "#67000d") +
#  theme_minimal()

plt <- ggplot(aldex_stats, aes(x = logFC, y = logP, alpha = logP,
                         label = Label, label2 = Id)) +
  geom_point(color = "#67000d") +
  theme_minimal()

ggplotly(plt) #, tooltip = c("label", "label2"))
```



```{r}
genome_of_interest = 'MAG10'
ggplot(data %>%
         as.data.frame %>%
         rownames_to_column(var = "Sample") %>%
         left_join(rownames_to_column(metadata), by = c("Sample" = "rowname")),
         aes_string(y = genome_of_interest, x = "Diet", fill = "Diet")) +
  geom_boxplot() +
  geom_beeswarm(cex=5) +
  theme_minimal() +
  scale_fill_manual(values = c(chow = "#00BFC4", HF = "#F8766D"))

kable(Tax %>%
  filter(user_genome == genome_of_interest))
```


# Functional Differences

```{r, load_modules}

kegg_modules= read_tsv(keggmodules_file,col_select = -1,show_col_types = FALSE)
module_names= kegg_modules %>% dplyr::select(c('module','module_name')) %>% distinct()


```

```{r, module_presence}

step_coverage_threshold= 0.8

module_step_coverage_matrix = pivot_wider(kegg_modules,  
                                   id_cols = genome,
                                   names_from = module,
                                   values_from = step_coverage
                                   ) %>%
                              column_to_rownames("genome") %>% as.matrix()

module_step_coverage_matrix = module_step_coverage_matrix[, colSums(module_step_coverage_matrix) > 0]


hatmap<- pheatmap(module_step_coverage_matrix, show_colnames = F)

module_presence_matrix = 1 * (module_step_coverage_matrix>step_coverage_threshold)

module_presence_matrix = module_presence_matrix[, colSums(module_presence_matrix) > 0]

hatmap
```

```{r, module_abundance}

# Sum of rel_ab for all species where a module is presence is equel to the matrix multiplication
stopifnot(dim(rel_ab)[2] == dim(module_presence_matrix)[1]  )
module_rel_ab <- as.matrix(rel_ab) %*% module_presence_matrix

```



```{r,heatmap_interactive2}


library(heatmaply)
d_tax= column_to_rownames(Tax,"user_genome")



annotations <- module_rel_ab
for (sample in rownames(annotations))
{
  for (module in colnames(annotations))
  {
    annotations[sample,module] = paste0("Diet: ",metadata[sample,"Diet"],
                       "\nPathway: ", module_names[module,"module_name"]
                       )
  }
}




heatmaply(module_rel_ab,
          custom_hovertext = annotations,
          showticklabels = c(FALSE,TRUE),
          row_side_colors = metadata$Diet,
          colors="RdBu"
          )


```


Hover over the heatmap to identify modules that are more abundant in the HF diet.



```{r, non_interactive_module_abundance}
# Non interactive plot

# make a dataframe to use to annotate the heatmap
annot_df <- data.frame(Sample = colnames(sig_data)) %>%
  left_join(rownames_to_column(metadata), by = c("Sample" = "rowname")) %>%
  column_to_rownames(var = "Sample") %>%
  dplyr::select(Diet)

# sort labels by sig_data order
heatmap_labels <- Tax %>%
  filter(user_genome %in% rownames(sig_data))
heatmap_labels <- heatmap_labels[order(match(heatmap_labels$user_genome, rownames(sig_data))), ]


pheatmap(module_rel_ab, show_colnames = F, annotation_row =dplyr::select(metadata,"Diet"))



```
